text/0803-type-ascription.md

```diff
@@ -172,24 +172,20 @@ lvalue position), then we don't have the soundness problem, but we do get the
 unexpected result that `&(x: T)` is not in fact a reference to `x`, but a
 reference to a temporary copy of `x`.
 
-The proposed solution is that type ascription expressions are rvalues, but
-taking a reference of such an expression is forbidden. I.e., type asciption is
-forbidden in the following contexts (where `<expr>` is a type ascription
-expression):
+The proposed solution is that type ascription expressions are lvalues. If the
```

Comment:
I find this phrasing kind of odd, though I think I agree with the intention.

It seems odd to say that `<expr>: T` is always an lvalue. For example, is `foo(): T` an *lvalue*? The only way I can make sense of that is if `<expr>: T` puts `<expr>` into a temporary (when it is an rvalue) and then is equivalent to referencing that temporary. 

My original thought was that `<expr>: T` would be an lvalue if `<expr>` is an lvalue, and an rvalue if `<expr>` is an rvalue. Moreover, there would be an additional rule that when `<expr>: T` appears in a ref context, the type of `<expr>` and `T` must match exactly (though, strictly speaking, I think it is ok to upcast if `<expr>` is an rvalue). 

That said, I think that the "temporary interpretation" I gave above and my interpretation behave the same way from an end-user point of view. For example, if you write `&(foo(): T)`, and you consider `foo(): T` to be an rvalue, then this will create a temporary with the enclosing temporary scope and store `foo()` in it. But if you take the first interpretation, it will also create a temporary and store `foo()` in it, and presumably the lifetime of that temporary would be the same (though we would probably want to adjust the temporary lifetime rules to "see through" `<expr>: T`).

Anyway, just for reference, let me define my terms a bit:

- lvalue: something that can be assigned to, e.g. `x`, `x.f`
- rvalue: something that cannot be assigned to, e.g. `x()`
- lvalue/ref context: something where you are taking the address, e.g. `<expr>` in `&<expr>` or in a match, etc.